### ISSUE TITLE
Feature file parsing code returns a hashtable instead of a list/array for single-column tables

### ISSUE THREAD

**Issue by fourpastmidnight** [pester/Pester#1121]
## 1. General summary of the issue

When the Gherkin file is parsed and a step definition table is encountered which contains only a single column, the `$table` variable's Key/Value pairs are incorrectly generated.

## 2. Describe Your Environment

Pester version : 4.4.2 C:\Program Files\WindowsPowerShell\Modules\Pester\4.4.2\Pester.psd1
PowerShell version : 5.1.17134.165
OS version : Microsoft Windows NT 10.0.17134.0

## 3. Expected Behavior

Given I have the following gherkin step and PowerShell step definition, I expected the following output:

### Gherkin Step

```gherkin
Given the module manifest has the following properties:
    # Remember, I can't use PropertyName because the text "Proper" causes an exception
    # as reported in issue #1119 
    | PropName      |
    | ModuleVersion |
    | GUID          |
    | Author        |
    | CompanyName   |
    | Copyright     |
    | Description   |
```

### PowerShell Step Definition
```powershell
Given "the module manifest has the following properties:?" {
    Param($tableRows)

    Write-Debug "Table has $($tableRows.Keys.Count) keys:"
    Write-Debug $($tableRows | Out-String)
}
```

### Expected Output

In SpecFlow, the Cucumber implementation for .NET, I would have accessed the table programmatically, as follows:

```csharp
foreach (var row in table.Rows)
{
    string[] propertyNames = row["PropName"]
}

// Or, using a for loop, treating the table as a 2-d array much like Cucumber in Ruby does:
var tableRow = table.Rows[0];
string[] propertyNames = new string[tableRow.Values.Count];
for (var i = 0; i < tableRow.Values.Count; ++i)
{
    var propertyNames[i] = tableRow[i];
}

// OR, SpecFlow has the ability to transform a table such that the columns become rows
// and the rows become columns:
var values = table.Transform()
```

Hmm, and with newer versions of SpecFlow (I've only recently returned to using it), I may even have been able to have a step definition method signature of:

```csharp
[Given("the module manifest contains the following properties:?")]
public void GivenTheModuleManifestContainsTheFollowingProperties(List<string> propertyNames)
{
    // Do something here
}
```

I expected that the `Given` step definition would produce a similar table or list of strings (or an array of strings) such that the step definition above would output the following:

```
Table has 1 keys:
Name        Value
----        -----
PropName    {ModuleVersion, GUID, Author, CompanyName, Copyright, Description}
```

and the equivalent in PowerShell to process the table would've been something like:

```powershell
Given "the module manifest has the following properties:?" {
    Param($tableRows)

    [string[]]$PropertyNames = @()

    for ($i = 0; $i -lt $tableRows.Count; $i++) {
        $PropertyNames[$i] = $tableRows[$i]
    }
}
# OR
Given "the module manifest has the following properties:?" {
    Param([string[]]$propertyNames)

    $Properties = $propertyNames
}
```

## 4.Current Behavior

Instead, from the same Gherkin Step and PowerShell step definition shown above, I received the following output:

```
Table has 48 keys:
Name    Version
----    -------
e
m
a
r
N
P       ModuleVersion
p
o
e
m
a
r
N
P       GUID
p
o
e
m
a
r
N
P       Author
p
o
e
m
a
r
N
P       CompanyName
p
o
e
m
a
r
N
P       Description
p
o
```

Looking at the `Name` column, it would appear the table header, `PropName` was split up character by character to represent Keys in the hash table. The only good thing about this is that the value portion is consistently on the upper-case `P` of the column header. I'm not sure why keys are ordered the way they are--they're not in order of the characters as they appear in the header, but they're not in character code order, either.

## 5. Possible Solution

Process tables more akin to how Cucumber for Ruby and SpecFlow for .NET process and represent them.

## 6. Context

This output is problematic because I was only expecting 6 rows but got 8x as many rows! Not only that, there are values for all 48 rows, and my `Then` condition was that these module manifest properties have a value (not any specific value, just a non-empty value). Well, all 48 rows passed the test. I thought my test was passing. In fact, it should have been failing! I got a false passing test! Those are the worst--they're worse than no test at all because they tell gave me a false sense of confidence that my code is working correctly.

So, how do I go about getting the correct values out of this table? I could _assume_ that the value will always be associated with the Key whose value is the first character of the column header text. But this is building in an assumption based on the (possible) incorrect behavior of the framework.

Assuming I haven't incorrectly used the `$table` parameter nor incorrectly specified the step definition (which are possibilities since I'm new to using Pester), then it's clear that the code which extracts a table from the Gherkin file does not handle single-column tables correctly.

**Comment by fourpastmidnight** [comment]
I may have been using the `$table` value incorrectly, I'm unsure. So if someone who's "in the know" could let me know, that would be great. I tried one more thing and got something _close_ to what I expected:

```powershell
Given "the module manifest has the following porperties:?" {
    Param($table)

    Write-Debug "Table has $($table.Values.Count) values."
    $PropertyNames = $table.Values | Out-String
    Write-Debug $PropertyNames
}

<#
Output:

DEBUG: Table has 48 values.
DEBUG: ModuleVersion
GUID
Author
CompanyName
Copyright
Description

    [+] Given the module manifest has the following properties:...
#>
```

**Comment by fourpastmidnight** [comment]
Even though I closed this, please note that it's not possible to use a single-column table as an array of strings until #1119 is resolved. Until then, you will  _most likely_ be the recipient of an `InvalidOperationException` due to Pester improperly parsing the single-columned table.

**Comment by nohwnd** [comment]
@Jaykul could you have a look please?

**Comment by Jaykul** [comment]
The original problem here was a minor bug, which will be fixed by @mikeclayton in #1127 -- at that point, this will be working **as designed**.

The edited post is asking for a new feature:

# Invoke-Gherkin should be able to convert Gherkin hashtables to other collection types

If you actually needed to write conversion loops the way you did in the example, I would agree we should do that within `Invoke-Gherkin`, but honestly the way to get a simple list from a Gherkin table is just to write your table with one column as you did, and then reference it like this:

```
Given ".*" { 
    param($Table)
    foreach ($Value in $Table.PropName) {
        <# ... #>
    }
}
```

Unlike other languages that don't unroll, this is **simple**, and doesn't seem to merit adding extra magic. PowerShell even supports casting strings to all sorts of things, so if you had numbers or whatever, we don't need to cast them for you, because you can just:

```
Given ".*" { 
    param($Table)
    foreach ([int]$Value in $Table.PropName) {
        <# ... #>
    }
}
```

**Comment by fourpastmidnight** [comment]
Ah, yes, when I was originally writing my tests, I was just getting back into PS. I’m traditionally a C# dev and need to constantly remind myself to think differently when coding PS.

Anyway, I agree, that’s really simple and doesn’t require any additional (complex) coding (magic) in Pester.

Sent from my Windows 10 phone
